tidy3d/components/geometry.py

```diff
@@ -3105,8 +3108,149 @@ def _surface_area(self, bounds: Bound) -> float:
         return area
 
 
+class CustomSurfaceMeshGeometry(Geometry, ABC):
+    """Custom surface geometry given by a triangle mesh, as in the STL file format.
+
+    Example
+    -------
+    >>> import stl
+    >>> stl_mesh = stl.mesh.Mesh(np.zeros(4), dtype=stl.mesh.Mesh.dtype)
+    >>> # tetrahedron
+    >>> stl_mesh.vectors = [
+    >>>     [[1,0,0],[0,1,0],[0,0,1]],
+    >>>     [[0,0,0],[0,0,1],[0,1,0]],
+    >>>     [[0,0,0],[1,0,0],[0,0,1]],
+    >>>     [[0,0,0],[0,1,0],[1,0,0]]
+    >>> ]
+    >>> stl_geom = CustomSurfaceMeshGeometry.from_stl(stl_mesh)
+    """
+
+    mesh_dataset: Optional[SurfaceMeshDataset] = pydantic.Field(
+        ...,
+        title="Surface mesh data",
+        description="Surface mesh data.",
+    )
+
+    @pydantic.validator("mesh_dataset", pre=True, always=True)
+    def _warn_if_none(cls, val: SurfaceMeshDataset) -> SurfaceMeshDataset:
+        """Warn if the DataArray fails to load."""
+        if isinstance(val, dict):
+            if any((v in DATA_ARRAY_MAP for _, v in val.items() if isinstance(v, str))):
+                log.warning("Loading 'mesh_dataset' without data.")
+                return None
+        return val
+
+    @classmethod
```

Comment:
let's add another `classmethod` called `from_numpy` that just accepts a numpy array, shaped `(N, 3, 3)`, storing the coordinates of the surface triangles and then converts them to the appropariate Dataset / DataArray? I think it would be nice to provide some more convenient method besides initializing the dataset/ data array